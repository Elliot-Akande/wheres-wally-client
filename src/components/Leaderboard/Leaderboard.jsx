import PropTypes from "prop-types";
import formatTime from "../../utils/formatTime";
import {
  Heading,
  Item,
  List,
  Name,
  Score,
  Status,
  StyledLeaderboard,
} from "./styles";

function Leaderboard({ data, loading }) {
  const getContent = () => {
    if (loading) {
      return <Status>Loading...</Status>;
    }

    if (data.scores.length === 0) {
      return <Status>No scores recorded.</Status>;
    }

    return (
      <List>
        {data.scores.map((entry) => (
          <Item key={entry._id}>
            <Name>{entry.name}</Name>
            <Score>{formatTime(entry.score)}</Score>
          </Item>
        ))}
      </List>
    );
  };

  return (
    <StyledLeaderboard>
      <Heading>Leaderboard</Heading>
      {getContent()}
    </StyledLeaderboard>
  );
}

Leaderboard.propTypes = {
  data: PropTypes.shape({
    scores: PropTypes.arrayOf(
      PropTypes.shape({
        _id: PropTypes.string,
        name: PropTypes.string,
        score: PropTypes.number,
      })
    ),
  }),
  loading: PropTypes.bool,
};

export default Leaderboard;
