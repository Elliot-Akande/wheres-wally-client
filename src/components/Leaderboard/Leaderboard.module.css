.container {
  width: 260px;
  padding: 16px;

  background-color: var(--neutral-500);
  border-radius: 8px;
}

.heading {
  margin: 0;
  padding-bottom: 16px;
}

.status {
  margin: 0;
}

.list {
  margin: 0;
  padding: 8px;
  height: 180px;

  display: flex;
  flex-direction: column;
  gap: 4px;

  overflow-y: scroll;

  list-style: none;
  background-color: var(--neutral-400);
  border-radius: 4px;
}

.item {
  height: max-content;

  display: flex;
  justify-content: space-between;
  gap: 8px;

  border-bottom: 1px solid var(--neutral-700);
}

.name {
  flex: 1;
  white-space: nowrap;
  overflow-x: clip;
  text-overflow: ellipsis;
}

.score {
  text-align: end;
}
