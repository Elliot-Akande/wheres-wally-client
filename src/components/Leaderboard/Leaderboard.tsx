import formatTime from "../../utils/formatTime";
import {
  Heading,
  Item,
  List,
  Name,
  Score,
  Status,
  StyledLeaderboard,
} from "./styles";

interface LeaderboardProps {
  data: {
    scores: Array<{
      _id: string;
      name: string;
      score: number;
    }>;
  } | null;
  loading: boolean;
}

function Leaderboard({ data, loading }: LeaderboardProps) {
  const getContent = () => {
    if (loading) {
      return <Status>Loading...</Status>;
    }

    if (!data) {
      return <Status>Data not present.</Status>;
    }

    if (data.scores.length === 0) {
      return <Status>No scores recorded.</Status>;
    }

    return (
      <List>
        {data.scores.map((entry) => (
          <Item key={entry._id}>
            <Name>{entry.name}</Name>
            <Score>{formatTime(entry.score)}</Score>
          </Item>
        ))}
      </List>
    );
  };

  return (
    <StyledLeaderboard>
      <Heading>Leaderboard</Heading>
      {getContent()}
    </StyledLeaderboard>
  );
}

export default Leaderboard;
