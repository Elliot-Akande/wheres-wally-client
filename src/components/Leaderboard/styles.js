import { styled } from "styled-components";

export const Status = styled.p`
  margin: 0;
`;

export const StyledLeaderboard = styled.div`
  width: 260px;
  padding: 16px;

  background-color: var(--neutral-500);
  border-radius: 8px;
`;

export const Heading = styled.h2`
  margin: 0;
  padding-bottom: 16px;
`;

export const List = styled.ul`
  margin: 0;
  padding: 8px;
  height: 180px;

  display: flex;
  flex-direction: column;
  gap: 4px;

  overflow-y: scroll;

  list-style: none;
  background-color: var(--neutral-400);
  border-radius: 4px;
`;

export const Item = styled.li`
  height: max-content;

  display: flex;
  justify-content: space-between;
  gap: 8px;

  border-bottom: 1px solid var(--neutral-700);
`;

export const Name = styled.span`
  flex: 1;
  white-space: nowrap;
  overflow-x: clip;
  text-overflow: ellipsis;
`;

export const Score = styled.span`
  text-align: end;
`;
