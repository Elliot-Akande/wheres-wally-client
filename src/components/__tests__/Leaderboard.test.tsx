import { render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import Leaderboard from "../Leaderboard/Leaderboard";

vi.mock("../../utils/formatTime.js", () => ({
  default: (time) => time,
}));

const data = {
  scores: [
    {
      _id: "3",
      name: "Anon",
      score: 32,
    },
    {
      _id: "1",
      name: "Big Gary",
      score: 100,
    },
  ],
};

describe("Leaderboard", () => {
  it("renders heading and list on valid leaderboard data", () => {
    const { container } = render(<Leaderboard data={data} loading={false} />);
    expect(container).toMatchSnapshot();
  });

  it("renders heading and loading message when loading data", () => {
    const { container } = render(<Leaderboard data={{}} loading={true} />);
    expect(container).toMatchSnapshot();
  });

  it("renders heading and empty scores message when no data after loading", () => {
    const { container } = render(
      <Leaderboard data={{ scores: [] }} loading={false} />
    );
    expect(container).toMatchSnapshot();
  });
});
