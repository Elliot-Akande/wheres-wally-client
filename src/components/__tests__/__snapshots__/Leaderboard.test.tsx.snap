// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Leaderboard > renders heading and empty scores message when no data after loading 1`] = `
<div>
  <div
    class="sc-guDLey gwwqSR"
  >
    <h2
      class="sc-dmyCSP dRcuXB"
    >
      Leaderboard
    </h2>
    <p
      class="sc-beySPh NxqXj"
    >
      No scores recorded.
    </p>
  </div>
</div>
`;

exports[`Leaderboard > renders heading and list on valid leaderboard data 1`] = `
<div>
  <div
    class="sc-guDLey gwwqSR"
  >
    <h2
      class="sc-dmyCSP dRcuXB"
    >
      Leaderboard
    </h2>
    <ul
      class="sc-hLQSwg bFhvaQ"
    >
      <li
        class="sc-eDLKkx bmiIbR"
      >
        <span
          class="sc-jTQCzO bXZCUx"
        >
          Anon
        </span>
        <span
          class="sc-gLLuof jXpbup"
        >
          32
        </span>
      </li>
      <li
        class="sc-eDLKkx bmiIbR"
      >
        <span
          class="sc-jTQCzO bXZCUx"
        >
          Big Gary
        </span>
        <span
          class="sc-gLLuof jXpbup"
        >
          100
        </span>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Leaderboard > renders heading and loading message when loading data 1`] = `
<div>
  <div
    class="sc-guDLey gwwqSR"
  >
    <h2
      class="sc-dmyCSP dRcuXB"
    >
      Leaderboard
    </h2>
    <p
      class="sc-beySPh NxqXj"
    >
      Loading...
    </p>
  </div>
</div>
`;
